pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java

```diff
@@ -38,6 +46,9 @@ public class LanguageVersionDiscoverer {
     public LanguageVersionDiscoverer(LanguageRegistry registry, LanguageVersion forcedVersion) {
         this.languageRegistry = registry;
         this.forcedVersion = forcedVersion;
+
+        // Add pattern to recognize POM. This can be overridden by patterns added later. POM defaults to XML if not loaded.
+        addLanguageFilePattern(LanguageFilePattern.ofRegex(Pattern.compile("(.*/)?pom\\.xml$"), "pom", "xml"));
```

Comment:
This should be done from the language module itself, the same way it registers versions and CPD / PMD capabilities.

Reply:
I guess so, but the order in which languages are added is not specified so it isn't clear which pattern would win in case of conflicts. So for now I think it is a good start.

Reply:
Alternatively we could drop the semantics of ordering, and say that if several patterns match, we log a warning and ignore the file. This is already what happens with the current file extension matching. But I find it intuitive and useful to have an ordering between patterns, because it allows a user to override the default patterns...

Reply:
Alternatively, we could try something different…

All the cases I've identified so far are suffixes, may that be:
* a standard language extension (ie: `.xml`)
* a standard file name (ie: `pom.xml`), which is the extreme case of a "suffix"
* a longer suffix (ie: SF flow files: `flow-meta.xml`)

I'd say that that we have multiple valid options:
* processing the file with all configured rules for all aplicable languages would be reasonable (ie: I don't need an additional rule to require pom files to [specify an XML encoding](https://docs.pmd-code.org/latest/pmd_rules_xml_bestpractices.html#missingencoding))
* alternatively, if we assume it's always file suffixes, we don't need a regex, but a string, and can decide on applying the most-specific (longest) one.

Reply:
> processing the file with all configured rules for all aplicable languages would be reasonable (ie: I don't need an additional rule to require pom files to [specify an XML encoding](https://docs.pmd-code.org/latest/pmd_rules_xml_bestpractices.html#missingencoding))

Tbh I don't think this should be done by applying several languages to the same file. I'd like to be able to say that the language POM is a dialect of the language XML. Then XML rules apply on a POM file because POM _is-a_ XML, but there is still exactly one language that gets picked, namely POM because it is the most specific. This means the file is only parsed once. It would also allow to share other things between XML dialects, eg parser and other language extensions. 

Assuming all .xml files are XML files (or some dialect of it, like POM for pom.xml) we can arbitrate conflicts using the specificity of the language. But the user should still be able to override the default mappings I think.

> Bonus points, if we assume this is always suffixes, we don't need a separate mechanism for normal language mapping and user-specified mappings, even if we decide to apply all matching languages.

I think it's a good thing to have a separate user-defined mapping. That way we don't need to modify the Language instances themselves at runtime. And the user can override mappings.

Reply:
> We would need to convert the extensions into LanguageFilePatterns

If you want me to make this change in this PR I will, but as shown here it works without, and that is an implementation detail.

> The correct order of the language modules is maybe reverse: if pom depends on xml, then xml is before pom, but we need to add the patterns of pom before xml)

That's why I think the more intuitive order is that the last pattern added wins. Intuitively it is just overriding the patterns that are already there. It also makes sense because users that add patterns on the CLI are appending patterns to this list, which overrides the existing patterns. 

> So for that PR, I would then return all matched languages from the ordered languageFilePatterns-list keeping the same order (and not just the first match). That way, we log a warning, if two assign-language patterns would match. In the end, only the first match will be used (just like it is now).

IMO the point of defining an ordering is to be able to override existing patterns (like saying my XML files are actually some dialect of XML). Overriding means, ignoring other patterns that matched and taking precedence over them. Therefore for this to be useful there will necessarily be other patterns that matched, and it is a normal occurence. We should log this in debug, as we log other file-language assignments, but I would not print a warning for what is the normal operation of things.